tests/cpp/test_spmat_csr.cc

```diff
@@ -202,44 +202,69 @@ TEST(SpmatTest, TestCSRGetRowData) {
 }
 
 template <typename IDX>
-void _TestCSRGetData() {
-  auto csr = CSR2<IDX>();
-  auto x = aten::CSRGetData(csr, 0, 0);
-  auto tx = aten::VecToIdArray(std::vector<IDX>({}), sizeof(IDX)*8, CTX);
+void _TestCSRGetData(DLContext ctx) {
+  auto csr = CSR2<IDX>(ctx);
+  // test get all data
+  auto x = aten::CSRGetAllData(csr, 0, 0);
+  auto tx = aten::VecToIdArray(std::vector<IDX>({}), sizeof(IDX)*8, ctx);
+  ASSERT_TRUE(ArrayEQ<IDX>(x, tx));
+  x = aten::CSRGetAllData(csr, 0, 2);
+  tx = aten::VecToIdArray(std::vector<IDX>({2, 5}), sizeof(IDX)*8, ctx);
+  ASSERT_TRUE(ArrayEQ<IDX>(x, tx));
+
+  // test get data
+  auto r = aten::VecToIdArray(std::vector<IDX>({0, 0, 0}), sizeof(IDX)*8, ctx);
+  auto c = aten::VecToIdArray(std::vector<IDX>({0, 1, 2}), sizeof(IDX)*8, ctx);
+  x = aten::CSRGetData(csr, r, c);
+  tx = aten::VecToIdArray(std::vector<IDX>({-1, 0, 2}), sizeof(IDX)*8, ctx);
   ASSERT_TRUE(ArrayEQ<IDX>(x, tx));
-  x = aten::CSRGetData(csr, 0, 2);
-  tx = aten::VecToIdArray(std::vector<IDX>({2, 5}), sizeof(IDX)*8, CTX);
+
+  // test get data on sorted
+  csr = aten::CSRSort(csr);
+  r = aten::VecToIdArray(std::vector<IDX>({0, 0, 0}), sizeof(IDX)*8, ctx);
+  c = aten::VecToIdArray(std::vector<IDX>({0, 1, 2}), sizeof(IDX)*8, ctx);
+  x = aten::CSRGetData(csr, r, c);
+  tx = aten::VecToIdArray(std::vector<IDX>({-1, 0, 2}), sizeof(IDX)*8, ctx);
   ASSERT_TRUE(ArrayEQ<IDX>(x, tx));
 
-  auto r = aten::VecToIdArray(std::vector<IDX>({0, 0, 0}), sizeof(IDX)*8, CTX);
-  auto c = aten::VecToIdArray(std::vector<IDX>({0, 1, 2}), sizeof(IDX)*8, CTX);
+  // test get data w/ broadcasting
+  r = aten::VecToIdArray(std::vector<IDX>({0}), sizeof(IDX)*8, ctx);
+  c = aten::VecToIdArray(std::vector<IDX>({0, 1, 2}), sizeof(IDX)*8, ctx);
   x = aten::CSRGetData(csr, r, c);
-  tx = aten::VecToIdArray(std::vector<IDX>({0, 2, 5}), sizeof(IDX)*8, CTX);
+  tx = aten::VecToIdArray(std::vector<IDX>({-1, 0, 2}), sizeof(IDX)*8, ctx);
   ASSERT_TRUE(ArrayEQ<IDX>(x, tx));
+
 }
 
-TEST(SpmatTest, TestCSRGetData) {
-  _TestCSRGetData<int32_t>();
-  _TestCSRGetData<int64_t>();
+TEST(SpmatTest, CSRGetData) {
+  _TestCSRGetData<int32_t>(CPU);
+  _TestCSRGetData<int64_t>(CPU);
+#ifdef DGL_USE_CUDA
+  _TestCSRGetData<int32_t>(GPU);
+#endif
 }
 
 template <typename IDX>
-void _TestCSRGetDataAndIndices() {
-  auto csr = CSR2<IDX>();
-  auto r = aten::VecToIdArray(std::vector<IDX>({0, 0, 0}), sizeof(IDX)*8, CTX);
-  auto c = aten::VecToIdArray(std::vector<IDX>({0, 1, 2}), sizeof(IDX)*8, CTX);
+void _TestCSRGetDataAndIndices(DLContext ctx) {
```

Comment:
Consider adding test for multigraphs.

Reply:
CSR2 is already a CSR with duplicate entries.